storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,22 +3571,401 @@ dberr_t fsp_sys_tablespace_validate()
+                               buf_pool_t::FIX_WAIT_READ);
+      if (!block)
+        continue;
+      if (!inodes->insert_seg(block->page.frame + TRX_RSEG +
+                              TRX_RSEG_FSEG_HEADER))
+        err= DB_CORRUPTION;
+      block->page.unfix();
+
+      /* Even after slow shutdown, there is a possiblity that
+      cached undo log can exist. So store the segment as used one */
+      for (trx_undo_t *undo= UT_LIST_GET_FIRST(rseg->undo_cached);
+           undo && err == DB_SUCCESS;
+           undo= UT_LIST_GET_NEXT(undo_list, undo))
+      {
+        block= buf_pool.page_fix(page_id_t{0, undo->hdr_page_no}, &err,
+                                 buf_pool_t::FIX_WAIT_READ);
+        if (!block)
+	  continue;
+	if (!inodes->insert_seg(block->page.frame + TRX_UNDO_SEG_HDR +
+                                TRX_UNDO_FSEG_HEADER))
+          err= DB_CORRUPTION;
+        block->page.unfix();
+      }
+    }
+  }
+
+  if (err == DB_SUCCESS)
+    err= fsp_table_inodes(inodes, mtr);
+  return err;
+}
+
+/** Free the extents, fragment page from the given inode
+@param page_no index node page number
+@param offset  index node offset within page
+@return error code */
+static void fseg_inode_free(uint32_t page_no, uint16_t offset)
+{
+  fil_space_t *space= fil_system.sys_space;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  sql_print_information("InnoDB: Freeing the unused segment page "
+                        UINT32PF " offset " UINT32PF, page_no,
+                        offset);
+  buf_block_t *iblock= buf_page_get_gen(page_id_t{0, page_no}, 0,
+                                        RW_X_LATCH, nullptr, BUF_GET,
+                                        &mtr);
+
+  DBUG_EXECUTE_IF("unused_undo_free_fail_4", iblock= nullptr;);
+  if (!iblock)
+  {
+    mtr.commit();
+    return;
+  }
+
+  fseg_inode_t *inode= iblock->page.frame + offset;
+  bool all_pages_freed= false;
+  while (!fseg_free_step_low(iblock, inode, space, &mtr, nullptr,
+                             &all_pages_freed))
+  {
+    DBUG_EXECUTE_IF("unused_undo_free_fail_5",
+                    mtr.commit();
+                    return;);
+    iblock->fix();
+    mtr.commit();
+
+    mtr.start();
+    mtr.x_lock_space(space);
+    iblock->page.lock.x_lock();
+    mtr.memo_push(iblock, MTR_MEMO_PAGE_X_FIX);
+  }
+
+  DBUG_EXECUTE_IF("unused_undo_free_fail_6", mtr.commit(); return;);
+  if (all_pages_freed)
+    /* Freeing completed: free the segment inode */
+    fsp_free_seg_inode(space, inode, iblock, &mtr);
+  mtr.commit();
+  return;
+}
+
+/** Free the unused segment
+@param boffset    offset of the list containing segment header
+@param space      system tablespace
+@param usedInodes Used system tablespace index node */
+static void fsp_free_unused_seg(inode_info *unused_inodes)
+{
+  for (auto it= unused_inodes->start();
+       it != unused_inodes->last(); it++)
+    fseg_inode_free((uint32_t)((*it) >> 32), (uint16_t)(*it));
+}
```

Comment:
It looks like this had better be a member function of `inode_info`. The `@param` comments are wrong. Would the following work in a member function?
```c++
  for (auto i : *this)
    fseg_inode_free(uint32_t(i >> 32), uint16_t(i));
```